build_tools/azure/install.sh

```diff
@@ -59,84 +48,29 @@ pre_python_environment_install() {
-            TO_INSTALL=""
-        fi
-
-        if [[ "$DISTRIB" == *"pypy"* ]]; then
-            TO_INSTALL="$TO_INSTALL pypy"
-        else
-            TO_INSTALL="$TO_INSTALL python=$PYTHON_VERSION"
-        fi
-
-        TO_INSTALL="$TO_INSTALL ccache pip blas[build=$BLAS]"
-
-        TO_INSTALL="$TO_INSTALL $(get_dep numpy $NUMPY_VERSION)"
-        TO_INSTALL="$TO_INSTALL $(get_dep scipy $SCIPY_VERSION)"
-        TO_INSTALL="$TO_INSTALL $(get_dep cython $CYTHON_VERSION)"
-        TO_INSTALL="$TO_INSTALL $(get_dep joblib $JOBLIB_VERSION)"
-        TO_INSTALL="$TO_INSTALL $(get_dep pandas $PANDAS_VERSION)"
-        TO_INSTALL="$TO_INSTALL $(get_dep pyamg $PYAMG_VERSION)"
-        TO_INSTALL="$TO_INSTALL $(get_dep Pillow $PILLOW_VERSION)"
-        TO_INSTALL="$TO_INSTALL $(get_dep matplotlib $MATPLOTLIB_VERSION)"
-
-        if [[ "$UNAMESTR" == "Darwin" ]] && [[ "$SKLEARN_TEST_NO_OPENMP" != "true" ]]; then
-                TO_INSTALL="$TO_INSTALL compilers llvm-openmp"
-        fi
-
-        make_conda $TO_INSTALL
+python_environment_install_and_activate() {
+    if [[ "$DISTRIB" == "conda"* ]]; then
+        conda update -n base conda -y
+        # pin conda-lock to latest released version (needs manual update from time to time)
+        conda install -c conda-forge conda-lock==1.0.5 -y
+        conda-lock install --name $VIRTUALENV $LOCK_FILE
+        source activate $VIRTUALENV
 
-    elif [[ "$DISTRIB" == "ubuntu" ]] || [[ "$DISTRIB" == "debian-32" ]]; then
+    elif [[ "$DISTRIB" == "ubuntu" || "$DISTRIB" == "debian-32" ]]; then
         python3 -m virtualenv --system-site-packages --python=python3 $VIRTUALENV
         source $VIRTUALENV/bin/activate
+        pip install -r "${LOCK_FILE}"
+    fi
 
-        python -m pip install $(get_dep cython $CYTHON_VERSION) \
-                $(get_dep joblib $JOBLIB_VERSION)
-
-    elif [[ "$DISTRIB" == "conda-pip-latest" ]]; then
-        # Since conda main channel usually lacks behind on the latest releases,
-        # we use pypi to test against the latest releases of the dependencies.
-        # conda is still used as a convenient way to install Python and pip.
-        make_conda "ccache python=$PYTHON_VERSION"
-        python -m pip install -U pip
-
-        python -m pip install pandas matplotlib scikit-image pyamg
-        # do not install dependencies for lightgbm since it requires scikit-learn.
-        python -m pip install "lightgbm>=3.0.0" --no-deps
-
-    elif [[ "$DISTRIB" == "conda-pip-scipy-dev" ]]; then
-        make_conda "ccache python=$PYTHON_VERSION"
-        python -m pip install -U pip
-        echo "Installing numpy and scipy master wheels"
+    if [[ "$DISTRIB" == "conda-pip-scipy-dev" ]]; then
+        echo "Installing development dependency wheels"
         dev_anaconda_url=https://pypi.anaconda.org/scipy-wheels-nightly/simple
         pip install --pre --upgrade --timeout=60 --extra-index $dev_anaconda_url numpy pandas scipy
-        pip install --pre cython
```

Comment:
> With this PR, are we still testing on the alpha builds for Cython 3.0?

Good point, I read this too fast and thought that because cython was not using `$dev_anaconda_url` the `--pre` was not doing anything ... but actually there are alpha Cython versions on PyPI.

> On the same note, conda-pip-scipy-dev is designed to test on nightly builds, should we pin at all fr this job?

I guess you still want to pin all the dependencies outside the dev dependencies, for example right now pytest and coverage versions are pinned in `build_tools/azure/pylatest_pip_scipy_dev_environment.yml` (same thing as in all the other builds)

Reply:
What do you think about removing the pins for NumPy, pandas, SciPy, joblib and pillow from `pylatest_pip_scipy_dev` and keeping all the other pins?

It would make the `pylatest_pip_scipy_dev` job less stable, but I think it is by design since it is testing the dev builds. This job only runs nightly, so it would not interrupt PRs.

Reply:
Goood point I did that in my latest commit. One of the potential downside is that this would not pin dependencies of numpy, pandas, scipy, joblib, and pillow.

Right now, I found the only dependencies which would not be pinned would be `python-dateutil` and `six` (`python-dateutil` is a dependency of `pandas`, six` is a dependency of `python-dateutil`). I have I added `python-dateutil` to the pinned dependencies.

If newer dependencies are introduced in the future we can add them